File: 542. 01 Matrix/01Matrix.c
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *columnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */

struct Pos {
    int index;
    int dist;
};


struct Pos queue[900000] = {0};
int        qi = 0;
int        qn = 0;


void testScore( int rows, int cols, int**result ) {
    
    int      dist = 0;
    int      index = 0;


    
    
    while( qi<qn ) {
        index = queue[qi].index;
        int row = index/cols;
        int col = index%cols;
        dist = queue[qi].dist;
        //printf( __FILE__ "(%d): %d/%d %d [%d][%d] %d\n", __LINE__, qi, qn, index, row, col, dist );


        qi++;
        if( qi==qn ) {
            qi = qn = 0;
        }

        if( result[row][col]>=0 && dist>result[row][col] )
            continue;
            
        //if( result[row][col]>=0 && dist-1>result[row][col] )
        //    continue;

        if( result[row][col] > dist || result[row][col]<0 )
            result[row][col] = dist;

        // UP
        if( row>0 ) {
            queue[qn].index = index-cols;
            queue[qn].dist = dist+1;
            qn++;
        }
        
        // RIGHT
        if( col+1<cols ) {
            queue[qn].index = index+1;
            queue[qn].dist = dist+1;
            qn++;
        }
        
        // DOWN
        if( row+1<rows ) {
            queue[qn].index = index+cols;
            queue[qn].dist = dist+1;
            qn++;
        }
        
        // LEFT
        if( col>0 ) {
            queue[qn].index = index-1;
            queue[qn].dist = dist+1;
            qn++;
        }
    }
}


int** updateMatrix(int** matrix, int matrixRowSize, int *matrixColSizes, int** columnSizes, int* returnSize) {
    int **result = (int**) calloc( matrixRowSize, sizeof(int*));

    *columnSizes = (int*) calloc( matrixRowSize, sizeof(int));
    *returnSize  = matrixRowSize;
    
    int row, col;
    for( row=0; row<matrixRowSize; row++  ) {
        result[row] = (int*) calloc( matrixColSizes[row], sizeof(int));

        (*columnSizes)[row] = matrixColSizes[row];
        
        for( col=0; col<matrixColSizes[row]; col++ ) {
            result[row][col] = -1;
        }
    }

    //#pragma omp parallel for
    for( row=0; row<matrixRowSize; row++ ) {
        for( col=0; col<matrixColSizes[row]; col++ ) {
            if( 0==matrix[row][col] ) {
                // 先把所有 0 起點放到 queue 裡面                
                result[row][col] = 0;
                queue[qn].index = row*matrixColSizes[row] + col;
                queue[qn].dist = 0;
                qn++;
                
            }
        }
    }
    
    // 從 queue 開始把迴圈跑完，分數就算完了
    testScore( matrixRowSize, matrixColSizes[0], result );
    return result;   
}

File: 542. 01 Matrix/01Matrix.go
type Pos struct {
    Row int
    Col int
    Dist  int
}


func updateMatrix(matrix [][]int) [][]int {
    que := make([]Pos, 0)
    
    result := make( [][]int, len(matrix))
    for y:=0; y<len(matrix); y++ {
        result[y] = make( []int, len(matrix[y]))
        for x:=0; x<len(matrix[y]); x++ {

            if matrix[y][x]==0 {
                result[y][x] = 0
                que = append( que, Pos{ y, x, 0})
            } else {
                result[y][x] = -1
            }
        }
    }
    
    for len(que)>0 {
        Row, Col, Dist := que[0].Row, que[0].Col, que[0].Dist
        que = que[1:] // move to next

        if result[Row][Col]>=0 && Dist>result[Row][Col] {
            continue
        }

        if result[Row][Col]>Dist || result[Row][Col]<0 {
            result[Row][Col] = Dist
        }

        // UP
        if Row>0 {
            que = append( que, Pos{ Row-1, Col, Dist+1 })
        }
        
        // RIGHT
        if Col+1<len(result[0]) {
            que = append( que, Pos{ Row, Col+1, Dist+1 })
        }
        
        // DOWN
        if Row+1<len(result) {
            que = append( que, Pos{ Row+1, Col, Dist+1 })
        }
        
        // LEFT
        if Col>0 {
            que = append( que, Pos{ Row, Col-1, Dist+1 })
        }
    }
    
    return result
}
